keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,652 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP extends the Job API to allow setting conditions under which an Indexed Job can be declared as succeeded.
+
+## Motivation
+
+There are cases where a batch workload requires an indexed job that want to care 
+only about leader indexes in determining the success or failure of a Job,
+for example MPI and PyTorch etc. This is currently not possible because the indexed job 
+is marked as Completed only if all indexes succeeded.
+
+Additionally, there are cases where a batch workload can be declared as succeeded,
+and continue running the rest of the pods if a number of pods succeed, regardless of the index.
+
+Some third-party frameworks have implemented success policy.
+
+- [Kubeflow Training Operator](https://www.kubeflow.org/docs/components/training)
+- [Flux Operator](https://flux-framework.org/flux-operator/)
+- [JobSet](https://github.com/kubernetes-sigs/jobset/)
+
+### Goals
+
+- allow to mark a job as a `Succeeded` according to a declared policy.
+
+### Non-Goals
+
+- Change the existing behavior of `NonIndexed` mode.
+- Change the existing behavior of `Indexed` mode without a SuccessPolicy.
+- Support SuccessPolicy for the job with `NonIndexed` mode: The SuccessPolicy can be theoretically supported for the job with `NonIndexed` mode. 
+However, we don't work on the job with `NonIndexed` mode in the first iteration since there aren't any effective use cases for the NonIndexed job.
+
+## Proposal
+
+We propose new policies under which a job can be declared as succeeded.
+Those policies can be modeled in the following:
+
+1. An indexed job completes if a set of [x, y, z...] indexes are successful.
+2. An indexed job completes if x or x% of indexes are successful.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a machine-learning researcher, I run an indexed job which leaders are launched
+as index=0 and workers are launched as index=1+.
+I want to care about only the leaders when the result of job is evaluated.
+
+In addition, we want to terminate running the rest of the pods if the leader index (index=0) is succeeded
+because the workers often don't have any ways to terminate themselves due to launching daemon processes like ssh-server.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+  - target:
+      succeededIndexes: "0"
+      whenTargetAchieved: terminate
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 2
+
+As a simulation researcher/engineer for fluid dynamics/biochemistry.
```

Comment:
As someone who was a former researcher, I don't know if I see the value in this user story. Was this coming from an actual usecase?

If I was a chemist, I'd be running a series of simulations to plot a trajectory (or whatever) and I would want to know which jobs failed so I can resubmit those. If I had a workflow and my jobs were successful but still were running, I worry that this may be confusing to a user.

Maybe @vsoch can overrule me if she thinks differently.

Reply:
@kannon92 I built this story based on https://github.com/kubernetes/enhancements/pull/4062#discussion_r1223434015.

Reply:
I see. I guess the main thing is about observability and how a user will know which jobs are continuing and what to do if they retry or fail. I'm thinking that someone would use this job in a workflow and they would probably want to know failures (or pods that are still running).

Reply:
Here is a great comment from @soltysh :https://github.com/kubernetes/enhancements/pull/4062/files#r1467756885

I think we can resolve observability by introducing a new condition, `Succeeded`, separate from completed.

@kannon92 WDYT?

Reply:
So I think this feature is valuable without the continue field. I think there is a lot of risks with continue and I'm not sure the usecase is so strong at the moment. If there is a user need for it, someone can always propose an enhancement for it.

But I think a condition could help so I'm happy with the conclusion if others are.

Reply:
If it is possible, I still want to include `whenTargetAchieve=continue` in this KEP since we have stories 2 and 3.
So, if anyone doesn't have strong objections, I want to add a `continue`  feature.

Reply:
We decided to leave the discussion `whenTargetAchived=continue` on beta graduation.
So, successPolicy doesn't support `continue` in the alpha stage.

Reply:
excellent, that simplifies a lot.

And I agree that this use case isn't very compeling.

Reply:
I marked a `continue` action as UNRESOLVED.

Reply:
Done.

Reply:
It makes sense.